### ISSUE TITLE
[REQUEST] An option for SUM gradient allreduce instead of MEAN

### ISSUE THREAD

**Issue by sfc-gh-lmerrick** [deepspeedai/DeepSpeed#7107]
**Is your feature request related to a problem? Please describe.**
In some learning problems, the correct allreduce of gradients across data-parallel workers is SUM rather than MEAN. For example, when doing contrastive learning for embedding models, it is common to allgather embeddings across all workers during the forward pass, compute a global loss, and then back-propagate, making the correct gradient reduction a SUM aggregation. Both PyTorch DDP and DeepSpeed currently appear to hard-code a `1/world_size` rescaling of gradients into the distributed backwards pass before executing a SUM allreduce, essentially hard-coding the backwards pass allreduce to be a MEAN operation.

To work around this hard-coded behavior, practitioners have to manually re-scale the loss value by `world_size` before back-propagating (e.g. as [this PyTorch community post suggests](https://discuss.pytorch.org/t/is-average-the-correct-way-for-the-gradient-in-distributeddataparallel-with-multi-nodes/34260/21)). Other workarounds include disabling or manually adjusting gradient clipping in proportion to the number of workers while simultaneously using a scale-invariant optimizer like Adam.

**Describe the solution you'd like**
I would like DeepSpeed to support for an additional configuration which would simply disable the `1/world_size` scaling, e.g. my config could look like `{"zero_optimization": {"stage": 1}, "gradient_allreduce_op": "sum"}`.

**Describe alternatives you've considered**
I have thought about trying to subclass `DeepSpeedZeroOptimizer` to modify its `average_tensor` method to not actually average anything but perform a SUM allreduce (I believe that would just involve gating [line 1122](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/stage_1_and_2.py#L1122) on a check that the config is the default value of "MEAN" instead of "SUM"). However, this would only apply to ZeRO stage1 and stage2, and I couldn't figure out a clean way to actually use my own subclass in my training loop.

**Additional context**
I chatted a bit with @stas00 about this possible feature, who mentioned this might be of interest to @tjruwase

**Comment by tjruwase** [comment]
@sfc-gh-lmerrick, this would be a useful feature. Are you able to provide a PR? I will be glad to work with you.

**Comment by sfc-gh-lmerrick** [comment]
I am happy to take a stab at a PR, though I'm not confident about handling the tricky case of Zero stage 3. Give me a couple days!

**Comment by tjruwase** [comment]
> , though I'm not confident about handling the tricky case of Zero stage 3

The first rule in taming ZeRO stage 3 is acknowledging its trickiness, so you are already on the right track :). 

Starting with ZeRO stage 1/2 is perfectly fine. Unfortunately, the code won't win any pageants. Here are few spots to note:
1. [line 1122](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/stage_1_and_2.py#L1122) (which you noted).
2. [here](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/stage_1_and_2.py#L993-L995).
3. [here](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/stage_1_and_2.py#L999-L1001).
4. [here](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/stage_1_and_2.py#L1003).
5. [here](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/stage_1_and_2.py#L1505-L1506).

The above code paths are mainly controlled by [reduce_scatter](https://github.com/deepspeedai/DeepSpeed/blob/b4177e45f9c9e1c49dc2efb88f85ad40e629d04e/deepspeed/runtime/zero/config.py#L28) option.

**Comment by sfc-gh-lmerrick** [comment]
Quick update: I have unfortunately been buried under a pile of urgent work items this past week and have yet to take a real stab at this, but I am hoping to get a chance in the next week or so to take a look!

**Comment by jinyouzhi** [comment]
Can I help? It looks like it's been pending for a while. @sfc-gh-lmerrick
